lib/Doctrine/ODM/MongoDB/Types/BinaryUuidType.php

```diff
@@ -0,0 +1,78 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Doctrine\ODM\MongoDB\Types;
+
+use Exception;
+use InvalidArgumentException;
+use MongoDB\BSON\Binary;
+use Symfony\Component\Uid\Uuid;
+
+use function get_debug_type;
+use function is_string;
+use function sprintf;
+
+class BinaryUuidType extends Type
+{
+    public function convertToDatabaseValue(mixed $value): ?Binary
+    {
+        return match (true) {
+            $value === null => null,
+            $value instanceof Binary => $value,
+            $value instanceof Uuid => new Binary($value->toBinary(), Binary::TYPE_UUID),
+            is_string($value) => new Binary(Uuid::fromString($value)->toBinary(), Binary::TYPE_UUID),
+            default => throw new InvalidArgumentException(sprintf('Invalid data type %s received for UUID', get_debug_type($value))),
+        };
+    }
+
+    public function convertToPHPValue(mixed $value): Uuid
+    {
+        if ($value instanceof Uuid) {
+            return $value;
+        }
+
+        if (! $value instanceof Binary) {
+            throw new Exception(sprintf('Invalid data of type "%s" received for Uuid', get_debug_type($value)));
+        }
+
+        if ($value->getType() !== Binary::TYPE_UUID) {
+            throw new Exception(sprintf('Invalid binary data of type %d received for Uuid', $value->getType()));
```

Comment:
Having a more specific exception make it catchable.
```suggestion
            throw MongoDBException::invalidValueForType(sprintf('Invalid binary data of type %d received for Uuid', $value->getType()));
```

Reply:
Replaced this with `InvalidArgumentException` which is used in all other places and types.

Reply:
I still believe we should have an exception in the `Doctrine\ODM\MongoDB\` namespace for this kind of error, which can occur at runtime when the data in the database is incorrect.

Reply:
Agree. Not sure if it should be a `MappingException` or if there should be a different class, but we should revisit this for a future version.

Reply:
Not a mapping error. Maybe `Doctrine\ODM\MongoDB\Types\ValueNotConvertible`. That would match `Doctrine\DBAL\Types\Exception\ValueNotConvertible`: https://github.com/doctrine/dbal/blob/2ce1218ef35cad74e488f7213df6dda701316569/src/Types/BinaryType.php#L39-L40